File: Data Structures & Algorithms/string-encode-and-decode/submission-0.java
class Solution {

    public String encode(List<String> strs) {
        StringBuilder stringBuilder = new StringBuilder();
        for (String s : strs) {
            stringBuilder.append(s.length()).append('#').append(s);
        }
        return stringBuilder.toString();
    }

    public List<String> decode(String str) {
        List<String> result = new ArrayList<>();
        int i = 0;
        while (i < str.length()) {
            int delimiter = str.indexOf('#', i);
            int length = Integer.parseInt(str.substring(i, delimiter));
            String curWord = str.substring(delimiter + 1, delimiter + 1 + length);
            result.add(curWord);
            i = delimiter + 1 + length;
        }
        return result;
    }
}

File: Data Structures & Algorithms/interleaving-string/submission-0.py
class Solution:
    def isInterleave(self, s1: str, s2: str, s3: str) -> bool:
        if len(s1) + len(s2) != len(s3):
            return False

        memo = {}
        
        def backtracking(i, j):
            k = i + j

            if k == len(s3):
                return True
            
            if (i, j) in memo:
                return memo[(i, j)]

            result = False
            
            if i < len(s1) and s1[i] == s3[k]:
                result = backtracking(i + 1, j)
            
            if not result and j < len(s2) and s2[j] == s3[k]:
                result = backtracking(i, j + 1)
            
            memo[(i, j)] = result
            return result
        return backtracking(0, 0)